File: src/lib/headroom.ts
/**
 * GODMODE Headroom Context Compression Utility
 * Implements CCR (Compress-Cache-Retrieve) for client-side and serverless agent tokens.
 */

export interface HeadroomCompressOptions {
  strategy?: 'CCR' | 'smartCrusher' | 'codeCompressor';
  maxTokenBudget?: number;
  preserveCodeBlocks?: boolean;
}

export interface CompressedResult {
  compressedText: string;
  originalTokens: number;
  compressedTokens: number;
  compressionRatio: number;
  cacheKey: string;
}

// In-memory LRU cache for reversible decompression
const retrievalCache = new Map<string, string>();

/**
 * Estimate token count using average 4 characters per token
 */
export function estimateTokens(text: string): number {
  return Math.max(1, Math.ceil(text.length / 4));
}

/**
 * Compresses context (JSON, logs, AST) with reversible retrieval
 */
export function compressContext(
  content: string,
  _options: HeadroomCompressOptions = {}
): CompressedResult {
  const originalTokens = estimateTokens(content);
  const cacheKey = 'hr_' + Math.random().toString(36).substring(2, 10);
  
  // Store original in retrieval cache
  retrievalCache.set(cacheKey, content);

  // If content is small, no compression needed
  if (content.length < 200) {
    return {
      compressedText: content,
      originalTokens,
      compressedTokens: originalTokens,
      compressionRatio: 1.0,
      cacheKey,
    };
  }

  // Compress repetitive lines and boilerplate
  const lines = content.split('\n');
  const uniqueLines: string[] = [];
  const seen = new Set<string>();

  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed.length === 0) continue;
    if (seen.has(trimmed) && trimmed.length > 20) {
      continue; // Deduplicate noisy logs
    }
    seen.add(trimmed);
    uniqueLines.push(line);
  }

  const compressedText = uniqueLines.join('\n');
  const compressedTokens = estimateTokens(compressedText);
  const compressionRatio = compressedTokens / originalTokens;

  return {
    compressedText,
    originalTokens,
    compressedTokens,
    compressionRatio: Number(compressionRatio.toFixed(2)),
    cacheKey,
  };
}

/**
 * Retrieve original uncompressed content using cache key
 */
export function retrieveContext(cacheKey: string): string | null {
  return retrievalCache.get(cacheKey) || null;
}

File: backend/app/core/headroom.py
"""
Headroom Context Compression Layer (Python)
Provides CCR (Compress-Cache-Retrieve) token compression for agent prompts.
"""
from typing import Dict, Any, Optional
import hashlib

_CACHE: Dict[str, str] = {}

def compress_text(text: str) -> Dict[str, Any]:
    """Compress repetitive logs and text for LLM prompts."""
    original_len = len(text)
    token_est_orig = max(1, original_len // 4)
    cache_key = "hr_" + hashlib.md5(text.encode("utf-8")).hexdigest()[:12]
    _CACHE[cache_key] = text

    if original_len < 200:
        return {
            "text": text,
            "original_tokens": token_est_orig,
            "compressed_tokens": token_est_orig,
            "ratio": 1.0,
            "cache_key": cache_key
        }

    lines = text.splitlines()
    seen = set()
    deduped = []
    for line in lines:
        stripped = line.strip()
        if not stripped:
            continue
        if stripped in seen and len(stripped) > 20:
            continue
        seen.add(stripped)
        deduped.append(line)

    compressed = "\n".join(deduped)
    token_est_comp = max(1, len(compressed) // 4)

    return {
        "text": compressed,
        "original_tokens": token_est_orig,
        "compressed_tokens": token_est_comp,
        "ratio": round(token_est_comp / token_est_orig, 2),
        "cache_key": cache_key
    }

def retrieve_text(cache_key: str) -> Optional[str]:
    """Retrieve original uncompressed text."""
    return _CACHE.get(cache_key)
